### PR TITLE
Refactor subheadings assessment

### LINKED ISSUE #1946: Improve keyphrase in subheadings assessment
These were the new boundaries as originally proposed were 30% minimum/75% maximum, see https://github.com/Yoast/YoastSEO.js/issues/1311

These new boundaries turned out to be too strict when we first tried to implement them in the morphology release (especially the 30% minimum).

Some ideas to improve these boundaries:
* Take the level of subheadings into account (e.g., minimum requirement only for h2, h3; deeper levels probably not as important)
* Only avoid over-optimization, i.e. only have an upper limit (e.g., the originally proposed 75%) and no lower limit.

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary

This PR can be summarized in the following changelog entry:

* Introduces a new Recalibration-specific scoring schema, where the minimal number of subheadings with keyphrase or a synonym is taken to be 30% and the maximal number 75%. 
* For Recalibration-specific analysis, limits analysis to H2 and H3-level subheadings, so that the texts with lots of low-level subheadings are not punished for not using keyphrase in many of them.

## Todo
- [ ] Revisit feedback strings to see if they are actionable enough -- This is pending.
- [ ] The assessment has adjusted scores in cornerstone assessor, double-check if these scores are also applicable to the recalibrated version
- [ ] Add testing instructions

## Relevant technical choices:

* The SEO cornerstone assessor has no configuration overrides for recalibration, it is already on par with the regular cornerstone strictness.
* Added recalibration variant for:
  - Assessment `calculateResult`.
  - Assessment `isApplicable`.
  - Research `getSubheadings`.
* Added `getSubheadings` variants for the top level (H2 and H3).

## Test instructions

This PR can be tested by following these steps:

- Go to the development tool: `cd examples/webpack`.
- Install packages: `yarn`.

### With recalibration

- Start the dev server with recalibration: `yarn start-recalibration`
- Add some paragraphs without subheadings and add a keyphrase, for example: `keyphrase`.
- Ensure you do not see a `Keyphrase in subheading` result under the `SEO` results. This is testing `isApplicable`.
- Add `<h4>Ignored</h4>` in your content. You could add something else in there, as long as it's not the keyphrase.
- Ensure there is still no result.
- Add `<h2>Subheading</h2>` in your content.
- Ensure you see a red bullet with: `Keyphrase in subheading: Use more keyphrases or synonyms in your higher level subheadings!`. This is testing the `no matches` scenario.
- Ensure the first link points to `https://yoa.st/33m`.
- Ensure the second link points to `https://yoa.st/33n`.
- Change the content of the `h2` to the keyphrase so it becomes `<h2>keyphrase</h2>`. Match the content with what you used as keyphrase!
- Ensure you see a green bullet with: `Keyphrase in subheading: Your higher level subheading reflects the topic of your copy. Good job!`. This is testing the `isOneOfOne` scenario.
- Add `<h3>subheading</h3>` to your content.
- Ensure that you see a green bullet with: `Keyphrase in subheading: 1 of your higher level subheadings reflects the topic of your copy. Good job!` This is testing the `hasGoodNumberOfMatches (singular)` scenario.
- Remove `<h3>subheading</h3>` again.
- Add `<h3>keyphrase</h3>` to your content.
- Ensure you see a red bullet with: `Keyphrase in subheading: More than 75% of your higher level subheadings reflect the topic of your copy. That's too much. Don't over-optimize!`. This is testing the `too many matches` scenario.
- Add another `<h2>Subheading</h2>` to your content.
- Ensure you see a green bullet with: `Keyphrase in subheading: 2 of your higher level subheadings reflect the topic of your copy. Good job!`. This is testing the `hasGoodNumberOfMatches (plural)` scenario.

### Without recalibration

- Start the dev server normally: `yarn start`.
- Add some paragraphs without subheadings and add a keyphrase, for example: `keyphrase`.
- Ensure you do not see a `Keyphrase in subheading` result under the `SEO` results. This is testing `isApplicable`.
- Add `<h2>Subheading</h2>` in your content.
- Ensure you see an orange bullet with: `Keyphrase in subheading: Use more keyphrases or synonyms in your subheadings!`. This is testing the `no matches` scenario.
- Activate `use cornerstone`.
- Ensure the bullet changes to red.
- Ensure the first link points to `https://yoa.st/33m`.
- Ensure the second link points to `https://yoa.st/33n`.
- Add `<h4>keyphrase</h4>` to your content.
- Ensure you see an orange bullet with: `Keyphrase in subheading: Your subheading reflects the topic of your copy. Good job!`. This is testing the `one match` scenario.
- Deactivate `use cornerstone`.
- Ensure the bullet change to green.
- Add another `<h4>keyphrase</h4>` to your content.
- Ensure you see a green bullet with: `Keyphrase in subheading: 2 (out of 3) subheadings reflect the topic of your copy. Good job!`. This is testing the `> 1 match` scenario.

Fixes #1946

### CODE REVIEW

File: src/stringProcessing/getSubheadings.js [hunk 1]
```diff
@@ -1,16 +1,35 @@
...
  * Gets all subheadings from the text and returns these in an array.
  *
  * @param {string} text The text to return the headings from.
- * @returns {Array} Matches of subheadings in the text, first key is everything including tags, second is the heading
- *                  level, third is the content of the subheading.
+ *
+ * @returns {Array<string[]>} Matches of subheadings in the text, first key is everything including tags,
+ *                            second is the heading level, third is the content of the subheading.
  */
 function getSubheadings( text ) {
```
Comment (hansjovis) [inline src/stringProcessing/getSubheadings.js]: Perhaps we can rename this into `getHeadings`?
Since this also gathers all `h1` headings as well.

File: src/assessments/seo/SubHeadingsKeywordAssessment.js [hunk 2]
```diff
@@ -62,7 +82,12 @@ export default class SubHeadingsKeywordAssessment extends Assessment {
 	 * @returns {boolean} True when there is at least one subheading.
 	 */
 	hasSubheadings( paper ) {
-		const subheadings = getSubheadings( paper.getText() );
+		let subheadings;
+		if ( process.env.YOAST_RECALIBRATION === "enabled" ) {
+			subheadings = getSubheadingsTopLevel( paper.getText() );
+		} else {
+			subheadings = getSubheadings( paper.getText() );
+		}
```
[the remark this answers is not in the document]
Comment (igorschoester) [reply src/assessments/seo/SubHeadingsKeywordAssessment.js]: Originally I left this as an explicit if statement because it is the beta and the tree shaking should remove the block that is never used. However, we do not apply this like that in a lot of other cases. Also, it is very possible that the tree shaking can shake a ternary statement perfectly.

**TL;DR** Will change! 😄

File: src/stringProcessing/getSubheadings.js (hunk 1 above)
Comment (igorschoester) [reply src/stringProcessing/getSubheadings.js]: Interesting, but after recalibration takes over this function will no longer be used. We might remove it then. So I do not think it's worth renaming now.

**Comment by manuelaugustin** [comment]
CR of latest commits & overall acceptance test: 👍 

I've made some small changes:
- I've changed the feedback strings so that they all refer to "higher level subheadings" instead of only "subheadings" or "H2 and H3 subheadings".
- I've changed the conditions so that:
  - If you have exactly one higher level subheading and that subheading contains the keyphrase you get the feedback `Keyphrase in subheading: Your higher level subheading reflects the topic of your copy. Good job!`.
  - If you have multiple higher level subheadings of which one contains the keyphrase (and the percentage is within the green score) you ge tthe feedback `Keyphrase in subheading: 1 of your higher level subheadings reflects the topic of your copy. Good job!`

Please CR my commits and acceptance test the specific functionality I changed.

**Comment by igorschoester** [comment]
Acceptance ✅
